editing/edit-features-with-feature-linked-annotation/src/main/java/com.esri.samples.edit_features_with_feature_linked_annotation/EditFeaturesWithFeatureLinkedAnnotationController.java

```diff
@@ -0,0 +1,310 @@
+   */
+  private void selectFeature(IdentifyLayerResult layerResult) {
+
+    try {
+      // if the selected feature is a polyline
+      if (selectedFeature.getGeometry().getGeometryType() == GeometryType.POLYLINE) {
+        // create a polyline builder from the selected feature
+        PolylineBuilder polylineBuilder = new PolylineBuilder((Polyline) selectedFeature.getGeometry());
+        // get a list of parts of the selected polyline
+        List<Part> parts = polylineBuilder.getParts();
+        parts.forEach(part -> {
+          // if the selected feature is a polyline with any part containing more than one segment
+          // (i.e. a curve)
+          if (part.getPointCount() > 2) {
+            selectedFeature = null;
+            // show message reminding user to select straight (single segment) polylines only
+            new Alert(Alert.AlertType.WARNING, "Select straight (single segment) polylines only.").show();
+            // return early, effectively disallowing selection of multi segmented polylines
+            return;
+          } else {
+            // select the polyline feature
+            ((FeatureLayer) layerResult.getLayerContent()).selectFeature(selectedFeature);
+            selectedFeatureIsPolyline = true;
+          }
+        });
+      }
+      // if the selected feature is a point, select the feature
+      else if (selectedFeature.getGeometry().getGeometryType() == GeometryType.POINT) {
+        ((FeatureLayer) layerResult.getLayerContent()).selectFeature(selectedFeature);
+        // open a dialog to edit the feature's attributes
+        showEditableAttributes(selectedFeature);
+      } else {
+        new Alert(Alert.AlertType.WARNING, "Feature of unexpected geometry type selected.").show();
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Create a dialog with text fields to allow editing of the given feature's 'AD_ADDRESS' and
+   * 'ST_STR_NAM' attributes.
+   */
+  private void showEditableAttributes(Feature selectedFeature) {
+
+    // create a dialog to edit the attributes of the selected feature
+    EditAttributesDialog editAttributesDialog = new EditAttributesDialog(selectedFeature);
+
+    // show the dialog and wait for the user response
+    editAttributesDialog.showAndWait();
+  }
+
+  /**
+   * Update the attributes of the selected feature.
+   */
+  private void updateAttributes(Feature selectedFeature) {
+
+    // update feature in the feature table
+    ListenableFuture<Void> editResult = selectedFeature.getFeatureTable().updateFeatureAsync(selectedFeature);
+    editResult.addDoneListener(() -> {
+              try {
+                if (editResult.isDone()) {
+                  editResult.get();
+                }
+              } catch (InterruptedException | ExecutionException e) {
+                new Alert(Alert.AlertType.ERROR, "Cannot update attributes: " + e.getCause().getMessage()).show();
+              }
+            }
+    );
```

Comment:
Nice work on this, there's only a couple of things we could improve here.
In the first line, were technically creating a Future object, which we then unpack with `.get()` to get the actual 'edit result' from inside. So Let's call this object `editResultFuture`
Since the lambda inside the `DoneListener` will only ever run if the future `isDone()`, there no need to check that again within the listener. You can just go ahead and do `.get()`.
Lastly, we don't generally display the whole error message (i.e. ` e.getCause().getMessage()` in the alerts) but just show something in plain english, describing what may have gone wrong. So let's update that message

```suggestion
    ListenableFuture<Void> editResultFuture = selectedFeature.getFeatureTable().updateFeatureAsync(selectedFeature);
    editResultFuture.addDoneListener(() -> {
              try {
                editResultFuture.get();
              } catch (InterruptedException | ExecutionException e) {
                new Alert(Alert.AlertType.ERROR, "Error updating attributes.").show();
              }
            }
    );
```